vllm/model_executor/layers/sampler.py

```diff
@@ -163,19 +163,22 @@ def _apply_min_tokens_penalty(
             start_idx += sampling_metadata.prompt_lens[i] - 1
 
         min_tokens = sampling_params.min_tokens
-        if min_tokens > 0:
+        eos_token_id = sampling_params.eos_token_id
+        stop_token_ids = sampling_params.stop_token_ids
+        if min_tokens > 0 and (eos_token_id is not None or stop_token_ids):
             seqs_to_penalize = []
-            for i, seq_id in enumerate(seq_ids):
+            for j, seq_id in enumerate(seq_ids):
                 seq_data = sampling_metadata.seq_data[seq_id]
                 if len(seq_data.output_token_ids) < min_tokens:
-                    seqs_to_penalize.append(i)
+                    seqs_to_penalize.append(j)
 
             if seqs_to_penalize:
                 # convert to the index into logits
-                seqs_to_penalize = [start_idx + i for i in seqs_to_penalize]
-                # use set() to remove any duplicates
-                token_ids_to_penalize = set(sampling_params.stop_token_ids +
-                                            [sampling_params.eos_token_id])
+                seqs_to_penalize = [start_idx + j for j in seqs_to_penalize]
+                if eos_token_id is None or eos_token_id in stop_token_ids:
+                    token_ids_to_penalize = stop_token_ids
+                else:
+                    token_ids_to_penalize = stop_token_ids + [eos_token_id]
```

Comment:
I think it may be better to convert `token_ids_to_penalize` to `set()` object, in order to remove any duplicates

Reply:
stop_token_ids may have duplicates itself, for examples: if pass stop_token_ids=[2, 2, 111] to sampling_param by user.

Reply:
@njhill @simon-mo

Reply:
@DefTruth it would probably be better to de-dup the stop_token_ids up-front when the sampling params are received, so that we're not creating a set here on every iteration. Let me make an update for that.

Reply:
> @DefTruth it would probably be better to de-dup the stop_token_ids up-front when the sampling params are received, so that we're not creating a set here on every iteration. Let me make an update for that.

sounds better ~

Reply:
@DefTruth OK I've pushed a commit for that. Thanks, I think this looks much better!